wallet-new/src/Cardano/Wallet/Kernel/Pending.hs

```diff
@@ -8,57 +8,116 @@ module Cardano.Wallet.Kernel.Pending (
+--
+-- NOTE: we select "our" output addresses from the transaction and pass it along to the data layer
+newPending :: ActiveWallet
+           -> HdAccountId
+           -> TxAux
+           -> Maybe TxMeta
+           -> IO (Either NewPendingError ())
+newPending w accountId tx mbMeta = runExceptT $ do
+    newTx w accountId tx mbMeta (NewPendingMissingKey accountId) $ \ourAddrs ->
+        update' ((walletPassive w) ^. wallets) $ NewPending accountId (InDb tx) ourAddrs
 
 -- | Submit new foreign transaction
 --
 -- A foreign transaction is a transaction that transfers funds from /another/
 -- wallet to this one.
-newForeign :: ActiveWallet -> HdAccountId -> TxAux -> TxMeta -> IO (Either NewForeignError ())
-newForeign ActiveWallet{..} accountId tx meta = do
-    res <- update' (walletPassive ^. wallets) $ NewForeign accountId (InDb tx)
-    case res of
-        Left e -> return (Left e)
-        Right () -> do
-            putTxMeta (walletPassive ^. walletMeta) meta
-            modifyMVar_ walletSubmission (return . addPending accountId (Pending.singleton tx))
-            return $ Right ()
+newForeign :: ActiveWallet
+           -> HdAccountId
+           -> TxAux
+           -> TxMeta
+           -> IO (Either NewForeignError ())
+newForeign w accountId tx meta = runExceptT $ do
+    newTx w accountId tx (Just meta) (NewForeignMissingKey accountId) $ \ourAddrs ->
+        update' ((walletPassive w) ^. wallets) $ NewForeign accountId (InDb tx) ourAddrs
+
+-- | Submit a new transaction
+--
+-- Will fail if the HdAccountId does not exist or if some inputs of the
+-- new transaction are not available for spending.
+--
+-- If the transaction is successfully added to the wallet state, transaction metadata
+-- is persisted and the submission layer is notified accordingly.
+--
+-- NOTE: we select "our" output addresses from the transaction and pass it along to the data layer
+newTx :: forall e. ActiveWallet
+      -> HdAccountId
+      -> TxAux
+      -> Maybe TxMeta
+      -> e                                  -- ^ error value to return on missing key
+      -> ([AddrWithId] -> IO (Either e ())) -- ^ the update to run, takes ourAddrs as arg
+      -> ExceptT e IO ()
+newTx ActiveWallet{..} accountId tx mbMeta missingKeyErr upd = do
+    -- run the update
+    esk <- getWalletAccountCreds
+    void $ liftIO $ upd (ourAddrs esk)
+
+    -- process transaction on success
+    liftIO $ do
+        putTxMeta' mbMeta
+        submitTx accountId tx
+        return ()
+    where
+        addrs = NE.toList $ map txOutAddress (_txOutputs . taTx $ tx)
+        wid   = WalletIdHdRnd (accountId ^. hdAccountIdParent)
+
+        getWalletAccountCreds :: ExceptT e IO EncryptedSecretKey
+        getWalletAccountCreds = do
+            mbKey <- Map.lookup wid . Map.fromList <$> liftIO (getWalletCredentials walletPassive)
+            case mbKey of
+                Nothing  -> throwError missingKeyErr
+                Just esk -> return esk
```

Comment:
I feel the logic here isn't quite right. This is related also to the problem I mentioned above about the need (or not) for this new `NewPendingMissingKey` constructor. When you are submitting a transaction here, you check for any addresses that belong to that particular account, and if you cannot find a key for that account, you have no choice but to throw an error. But I feel that this is not quite the right way up. What if we submit a transaction to this account but it has an address that belongs to another account? Mind, I'm not saying that it's necessarily a _problem_ to recognize only addresses belonging to this one, but it would make sense to just recognize addresses belonging to _any_ account right? And if we did that, that would just mean iterating over the keys that we get from `getWalletCredentials` (just like you do during prefiltering), and then this error would magically vanish.

Reply:
clever solution, thanks! that cleans it up nicely...